### PR TITLE
refactor(migrations): migrate package files

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Class based migration of `packageFiles`

## Context
- #11459 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by RahulGautamSingh** [comment]
@rarkins Do we need to handle nested `packageFiles` or nested `packageRules` inside of `packageFiles` ?

File: lib/config/presets/bitbucket-server/index.ts [hunk 1]
```diff
@@ -32,7 +32,7 @@ export async function fetchJSONFile(
   try {
     res = await http.getJson(url);
   } catch (err) {
-    // istanbul ignore if: not testable with nock
+    // istanbul ignore if: not testable with mock
```
Comment (JamieMagee) [inline lib/config/presets/bitbucket-server/index.ts:35]: unrelated change?

File: lib/config/migrations/custom/package-files-migration.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  override readonly deprecated = true;
+  override readonly propertyName = 'packageFiles';
+
+  override run(value: unknown): void {
+    const packageRules: PackageRule[] = this.get('packageRules') ?? [];
+    if (is.array(value)) {
+      const fileList = [];
+      for (const packageFile of value) {
+        if (is.object(packageFile) && !is.array(packageFile)) {
+          fileList.push((packageFile as any).packageFile);
```
Comment (JamieMagee) [inline lib/config/migrations/custom/package-files-migration.ts:15]:
```suggestion
    if (is.array(value)) {
      const fileList: string[] = [];
      for (const packageFile of value) {
        if (
          is.nonEmptyObject(packageFile) &&
          'packageFile' in packageFile &&
          is.string(packageFile['packageFile'])
        ) {
          fileList.push(packageFile.['packageFile']);
```

File: lib/config/migrations/custom/package-files-migration.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+          if (Object.keys(packageFile).length > 1) {
+            packageRules.push({
+              ...packageFile,
+            });
+          }
+        } else {
+          fileList.push(packageFile);
+        }
```
Comment (JamieMagee) [inline lib/config/migrations/custom/package-files-migration.ts:23]:
```suggestion
        } else if (is.array<string>(packageFile)) {
          fileList.push(...packageFile);
        } else if (is.string(packageFile)) {
          fileList.push(packageFile);
        }
```

File: lib/config/presets/bitbucket-server/index.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/config/presets/bitbucket-server/index.ts:35]: Yes, found it during debugging. Should it be done in a new PR?

**Comment by RahulGautamSingh** [comment]
Verified Working here: https://github.com/RahulGautamSingh-testing/docker-helm/pull/5/files

File: lib/config/migrations/custom/package-files-migration.ts (hunk 3 above)
Comment (RahulGautamSingh) [reply lib/config/migrations/custom/package-files-migration.ts]: Will `packageFile` ever be an array?

File: lib/config/presets/bitbucket-server/index.ts (hunk 1 above)
Comment (viceice) [reply lib/config/presets/bitbucket-server/index.ts:35]: it's intentionally `nock`, so please revert

Comment (RahulGautamSingh) [reply lib/config/presets/bitbucket-server/index.ts:35]: My bad

**Review by JamieMagee — DISMISSED** [review]

File: lib/config/migrations/custom/package-files-migration.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+          'packageFile' in packageFile &&
+          is.string(packageFile['packageFile'])
+        ) {
+          fileList.push(packageFile['packageFile']);
+          if (Object.keys(packageFile).length > 1) {
+            packageRules.push({
+              ...packageFile,
+            });
+          }
+        } else if (is.array<string>(packageFile)) {
```
Comment (viceice) [inline lib/config/migrations/custom/package-files-migration.ts:25]:
```suggestion
        } else if (is.array(packageFile, is.string)) {
```
seems more safe

File: lib/config/migrations/custom/package-files-migration.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    const packageRules: PackageRule[] = this.get('packageRules') ?? [];
+    if (is.array(value)) {
+      const fileList: string[] = [];
+      for (const packageFile of value) {
+        if (
+          is.nonEmptyObject(packageFile) &&
+          'packageFile' in packageFile &&
+          is.string(packageFile['packageFile'])
+        ) {
+          fileList.push(packageFile['packageFile']);
```
Comment (viceice) [inline lib/config/migrations/custom/package-files-migration.ts:19]:
```suggestion
          is.string(packageFile.packageFile)
        ) {
          fileList.push(packageFile.packageFile);
```
looks cleaner

**Review by viceice — COMMENTED** [review]
🤷‍♂️😉

**Review by JamieMagee — DISMISSED** [review]

**Review by JamieMagee — DISMISSED** [review]
Approved, but will give @viceice a chance for one more review

**Review by JamieMagee — DISMISSED** [review]

File: lib/config/migration.ts [hunk 6]
```diff
@@ -208,8 +175,10 @@ export function migrateConfig(config: RenovateConfig): MigratedConfig {
           logger.debug('Flattening nested packageRules');
           // merge each subrule and add to the parent list
           for (const subrule of packageRule.packageRules) {
-            // TODO: fix types #7154
-            const combinedRule = mergeChildConfig(packageRule, subrule as any);
+            const combinedRule = mergeChildConfig(
```
Comment (viceice) [inline lib/config/migration.ts:178]: please readd the Todo comment, it's still valid

**Review by viceice — COMMENTED** [review]
just a small thing

**Review by viceice — APPROVED** [review]
